parallel-install/pkg/git/clone.go

```diff
@@ -23,46 +25,65 @@ func (rc *remoteRepoCloner) Clone(url, path string, autoCheckout bool) (*git.Rep
 	})
 }
 
-var defaultCloner repoCloner = &remoteRepoCloner{}
+type RepoCloner interface {
+	Clone(dstPath string) error
+	Checkout(rev string) error
+	resolveRevision(rev string) (*plumbing.Hash, error)
+}
 
-// CloneRepo clones the repository in the given URL to the given dstPath and checks out the given revision.
-// revision can be 'main', a release version (e.g. 1.4.1), a commit hash (e.g. 34edf09a) or a PR (e.g. PR-9486).
-func CloneRepo(url, dstPath, rev string) error {
-	rev, err := ResolveRevision(url, rev)
+type Repo struct {
```

Comment:
Why introduce this type and make it exported? Exported members change the package's public API.

Reply:
I make it private, but my idea was to make the cloneRepo function more readable through encapsulation. Also we might potentially want to have more control over the cloning behavior in the future? The repo class could be a good boundary interface then

Reply:
In my opinion, this only makes the package usage more complex and adds more boilerplate without giving any benefits.
It would help thinking about it as package and its usage. It is more simple, readable and idiomatic to do (3 lines vs 7):
```go
if err := git.CloneRepo(url, path, rev); err != nil {
    return err
}
```
instead of:
```go
repo := git.NewRepo(url)
if err := repo.Clone(path); err != nil {
    return err
}
if err := repo.Checkout(rev); err != nil {
    return err
}
```

It is a harder to use API and more error prone. For example, what happens if I call `repo.Checkout()` before `repo.Clone()`? All these wrong API usage problems (the most common usually) go away automatically with a compact and clean API.

Regarding the control, we actually loose some as the previous potential misuse shows. We no longer control the order of the logic for example.